### PR TITLE
Add new syntax error for when command references variable that doesn't exist

### PR DESCRIPTION
**Pull request by scottfrazer**
This WDL file:

```
task a {
  Int x
  command { ./script ${x+y} }
}

workflow w {
  call a
}
```

Gives this syntax error:

```
ERROR: Variable does not reference any declaration in the task (line 3, col 26):

  command { ./script ${x+y} }
                         ^

Task defined here (line 1, col 6):

task a {
     ^
```

This WDL:

```
task a {
  Int x
  Int y = x + z
  command { ./script ${x} }
}

workflow w {
  call a {input: x=5}
}
```

Will produce this error:

```
ERROR: Variable does not reference any declaration in the task (line 3, col 15):

  Int y = x + z
              ^

Declaration starts here (line 3, col 7):

  Int y = x + z
      ^
```

### CODE REVIEW

**Comment by scottfrazer** [comment]
WHEEL: @mcovarr @geoffjentry

**Comment by scottfrazer** [comment]
@yfarjoun this one's for you :)

**Comment by yfarjoun** [comment]
YEY!

**Comment by cjllanwarne** [comment]
Does PyWDL have an equivalent change upcoming?

**Comment by scottfrazer** [comment]
Yeah, PyWDL is missing a lot of these higher-level syntax checks actually

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 1]
```diff
@@ -256,5 +256,17 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
...
   }
+  it should "detect when expressions within a command block don't reference actual variables" in {
+    expectError(
+      """task a {
+        |  Int x
+        |  command { ./script ${x+y} }
+        |}
+        |
+        |workflow w {
+        |  call a
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: From a test-pedantry point of view, this could just be throwing an error because "x" is never set. To be exact, make sure you say `call a {input: x=5 }` or equivalent

Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Actually, does assigning to a variable that's never declared also constitute a syntax error?

Comment (scottfrazer) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: I believe `call a {input: x=5, y=5 }` ought to return an syntax error in this case.  If that syntax error is not already present I'll add it.

**Comment by cjllanwarne** [comment]
This checks for unfound declarations in the command text. Does it also validate missing declarations in other declarations? e.g.
```
task X {
  Int x
  Int z = x + y
}
```

**Comment by scottfrazer** [comment]
@cjllanwarne I'll check on that too.  The algorithm should be the same to check those expressions so I'll extend it to also cover those expressions.

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala (hunk 1 above)
Comment (scottfrazer) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: I just confirmed that there is indeed a syntax error and test covering this already.

**Comment by scottfrazer** [comment]
@cjllanwarne I added the same kind of syntax check for declarations in tasks... see the description for what that error message looks like.

File: src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala [hunk 2]
```diff
@@ -204,4 +204,39 @@ case class WdlSyntaxErrorFormatter(terminalMap: Map[Terminal, WdlSource]) extend
...
+        |
+        |${pointToSource(variable)}
+        |
+        |Declaration starts here (line ${declarationName.getLine}, col ${declarationName.getColumn}):
+        |
+        |${pointToSource(declarationName)}
+     """.stripMargin
+  }
+
+  def variableDeclaredMultipleTimes(first: Terminal, second: Terminal) = {
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: Ooh, are we also covered for DSDEEPB-2531 now?

Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: Looking at the ticket, there's 3 test cases. I think this covers the first?

Comment (scottfrazer) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: I will look

File: src/main/scala/wdl4s/Task.scala [hunk 3]
```diff
@@ -16,15 +16,48 @@ import scala.util.Try
...
+        if (!declarations.map(_.name).contains(variable.getSourceString)) {
+          throw new SyntaxError(wdlSyntaxErrorFormatter.commandExpressionContainsInvalidVariableReference(nameTerminal, variable))
+        }
+      }
+    }
+
+    def findDeclarationAsts(decl: Declaration) =
+      ast.findAsts(AstNodeName.Declaration) collect { case a: Ast if a.getAttribute("name").sourceString == decl.name => a }
+
+    declarations.foldLeft(Seq.empty[Declaration]) { (acc, cur) =>
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/Task.scala]: This foldLeft function is a little unwieldy in-line. Can you break it out into a separate def?

File: src/main/scala/wdl4s/Task.scala [hunk 4]
```diff
@@ -30,6 +30,38 @@ object Task {
...
+      cur.expression match {
+        case Some(e) => e.ast.findTerminals("identifier") foreach {variable =>
+          if (!acc.map(_.name).contains(variable.getSourceString)) {
+            // .head below because we are assuming if you have a Declaration object that it must have come from a Declaration AST
+            val declarationAst = findDeclarationAsts(cur).head
+            throw new SyntaxError(wdlSyntaxErrorFormatter.declarationContainsInvalidVariableReference(declarationAst.getAttribute("name").asInstanceOf[Terminal], variable))
+          }
+        }
+        case None =>
+      }
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: converting the `match` to a `foreach` would eliminate the pattern match on `Some` and the need to have a `None` cass

File: src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala (hunk 2 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: I didn't see a test for this

File: src/main/scala/wdl4s/Task.scala [hunk 5]
```diff
@@ -30,6 +30,38 @@ object Task {
       case x: Ast => ParameterCommandPart(x, wdlSyntaxErrorFormatter)
     }
 
+    commandTemplate collect { case x: ParameterCommandPart => x } foreach { p =>
+      p.expression.ast.findTerminals("identifier") foreach {variable =>
```
Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: this `findTerminals("identifier")` is too permissive.  It also catches function names... I need to fix that.

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 6]
```diff
@@ -260,9 +260,6 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
     expectError(
       """task a {
         |  command { ./script }
-        |  output {
-        |    Int x = "bad value"
-        |  }
```
Comment (mcovarr) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Why the change to what appears to be an unrelated test?

File: src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala [hunk 7]
```diff
@@ -207,22 +207,57 @@ case class WdlSyntaxErrorFormatter(terminalMap: Map[Terminal, WdlSource]) extend
...
+    s"""ERROR: Variable '${first.getSourceString}' is declared more than once.
+        |
+        |First occurrence (line ${first.getLine}, col ${first.getColumn}):
+        |
+        |${pointToSource(first)}
+        |
+        |Second occurence (line ${second.getLine}, col ${second.getColumn}):
```
Comment (mcovarr) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: At least one of the spellings of the word after 'First' and 'Second' is incorrect.  :smile:

File: src/main/scala/wdl4s/Task.scala [hunk 8]
```diff
@@ -30,6 +30,38 @@ object Task {
...
+    def findDeclarationAsts(decl: Declaration) =
+      ast.findAsts(AstNodeName.Declaration) collect { case a: Ast if a.getAttribute("name").sourceString == decl.name => a }
+
+    declarations.foldLeft(Seq.empty[Declaration]) { (acc, cur) =>
+      findDeclarationAsts(cur) match {
+        case declAsts if declAsts.size > 1 =>
+          val declNameTerminals = declAsts.map(_.getAttribute("name").asInstanceOf[Terminal])
+          throw new SyntaxError(wdlSyntaxErrorFormatter.variableDeclaredMultipleTimes(declNameTerminals(0), declNameTerminals(1)))
+        case _ =>
+      }
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: could be a `collect` to get rid of the `case _` line

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala (hunk 6 above)
Comment (scottfrazer) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: I want to keep these WDL files to only contain the exact syntax error that they're actually testing for and nothing else.  In this case, the output section is not needed to test the `meta` section.  Worse than that though is that the output section itself contains a syntax error so it's actually masking the real syntax error of this test.  `expectError` really should probably match the error it gets back against a string so you know you got the right error.  I can file a ticket for that, because I don't want to go down too much of a rabbit hole

File: src/main/scala/wdl4s/Task.scala (hunk 8 above)
Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: hmmm I'm not sure how exactly...

File: src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala (hunk 2 above)
Comment (scottfrazer) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: I'll add it

File: src/main/scala/wdl4s/Task.scala (hunk 8 above)
Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: Maybe it's better as an if statement?

Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: sure, the collect I had in mind didn't actually make sense.  :smile:

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 9]
```diff
@@ -285,5 +282,77 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
...
+      """task inputOutputOops {
+        |  Int c = 5
+        |  command { echo 5 }
+        |  output {
+        |    Int c = 10
+        |  }
+        |}
+        |workflow c { call inputOutputOops }
+      """.stripMargin)
+  }
```
Comment (scottfrazer) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: @cjllanwarne there's your DSDEEPB-2531 WDLs ^^^

**Comment by scottfrazer** [comment]
@mcovarr @cjllanwarne I pushed all code review changes

File: src/main/scala/wdl4s/Task.scala [hunk 10]
```diff
@@ -30,7 +31,87 @@ object Task {
...
+    val variablesReferencedInCommand = for {
+      param <- commandTemplate collect { case x: ParameterCommandPart => x }
+      variable <- param.expression.variableReferences
+    } yield variable
+
+    variablesReferencedInCommand foreach { variable =>
+      if (!declarations.map(_.name).contains(variable.getSourceString)) {
+        throw new SyntaxError(wdlSyntaxErrorFormatter.commandExpressionContainsInvalidVariableReference(taskNameTerminal, variable))
+      }
+    }
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: I think this would be better as `Validation`s that accumulate errors rather than failing fast on the first problem

File: src/main/scala/wdl4s/Task.scala [hunk 11]
```diff
@@ -30,7 +31,87 @@ object Task {
...
+      if (declarationAstsWithSameName.nonEmpty)
+        throw new SyntaxError(wdlSyntaxErrorFormatter.variableDeclaredMultipleTimes(declNameTerminals(0), outputNameTerminals(0)))
+
+      current.expression.variableReferences foreach { variable =>
+        val varName = variable.getSourceString
+        if (!accumulated.map(_.name).contains(varName) && !declarations.map(_.name).contains(varName)) {
+          throw new SyntaxError(wdlSyntaxErrorFormatter.declarationContainsInvalidVariableReference(
+            taskOutputAstsWithSameName.head.getAttribute("var").asInstanceOf[Terminal],
+            variable
+          ))
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: More fast failing here

File: src/main/scala/wdl4s/AstTools.scala [hunk 12]
```diff
@@ -13,10 +14,10 @@ import scala.collection.JavaConverters._
 object AstTools {
   implicit class EnhancedAstNode(val astNode: AstNode) extends AnyVal {
     def findAsts(name: String): Seq[Ast] = AstTools.findAsts(astNode, name)
-    def findAstsWithTrail(name: String, trail: Seq[AstNode] = Seq.empty): Map[Ast, Seq[AstNode]] = {
+    def findAstsWithTrail(name: String, trail: Seq[AstNode] = Seq.empty): Map[Ast, Seq[AstNode]] =
       AstTools.findAstsWithTrail(astNode, name, trail)
-    }
-    def findTerminals(): Seq[Terminal] = AstTools.findTerminals(astNode)
+    def findTerminalsWithTrail(terminalType: String, trail: Seq[AstNode] = Seq.empty): Map[Terminal, Seq[AstNode]] =
+      AstTools.findTerminalsWithTrail(astNode, terminalType, trail)
```
Comment (mcovarr) [inline src/main/scala/wdl4s/AstTools.scala]: I believe our standards say to use curlies if it's not a one liner

File: src/main/scala/wdl4s/Task.scala [hunk 13]
```diff
@@ -30,7 +31,87 @@ object Task {
...
+    }
+
+    /**
+      * Ensures that the current declaration doesn't have a name conflict with another declaration
+      * and that the expression for the current declaration only has valid variable references in it
+      *
+      * @param accumulated The declarations that come lexically before 'current'
+      * @param current The declaration being validated
+      */
+    def validateDeclaration(accumulated: Seq[Declaration], current: Declaration): Seq[Declaration] = {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: and the way this is used will also result in fast failing on the first bad declaration

File: src/main/scala/wdl4s/Task.scala (hunk 10 above)
Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: What do you think now?

Comment (mcovarr) [inline src/main/scala/wdl4s/Task.scala]: Much better, thanks!  I would suggesting adding a note as to why converting this to `Validation`s is not straightforward.  I'm also not sure if there's any value in putting the errors in a `SyntaxError` box; these could perhaps just be `String`s.

Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: Okay, will do

Comment (scottfrazer) [inline src/main/scala/wdl4s/Task.scala]: Updated

File: src/main/scala/wdl4s/AstTools.scala [hunk 14]
```diff
@@ -16,7 +17,9 @@ object AstTools {
     def findAstsWithTrail(name: String, trail: Seq[AstNode] = Seq.empty): Map[Ast, Seq[AstNode]] = {
       AstTools.findAstsWithTrail(astNode, name, trail)
     }
-    def findTerminals(): Seq[Terminal] = AstTools.findTerminals(astNode)
+    def findTerminalsWithTrail(terminalType: String, trail: Seq[AstNode] = Seq.empty): Map[Terminal, Seq[AstNode]] = {
```
Comment (geoffjentry) [inline src/main/scala/wdl4s/AstTools.scala]: These aliases were originally just to minimize the impact of some changes I made a while back. Since this is a new one it might as well be fully and completely be in here instead of a passthrough.

**Comment by mcovarr** [comment]
:+1:

File: src/main/scala/wdl4s/AstTools.scala (hunk 14 above)
Comment (scottfrazer) [inline src/main/scala/wdl4s/AstTools.scala]: Okay it's moved

**Comment by geoffjentry** [comment]
Glam

**Comment by geoffjentry** [comment]
Ugh. Lgtm
